src/parser/scalar.js

```diff
@@ -79,6 +86,65 @@ module.exports = {
     return text.replace(removementRegExp, "\n");
   },
 
+  /**
+   * Check indentation level of heredoc in text, if mismatch, raiseError
+   * @param {string} text
+   * @param {number} indentation
+   * @param {boolean} indentation_uses_spaces
+   * @param {boolean} first_encaps_node if it is behind a variable, the first N spaces should not be removed
+   */
+  check_heredoc_indentation_level: function(
+    text,
+    indentation,
+    indentation_uses_spaces,
+    first_encaps_node
+  ) {
+    const textSize = text.length;
+    let offset = 0;
+    let leadingWhitespaceCharCount = 0;
+    /**
+     * @var inCoutingState {boolean} reset to true after a new line
+     */
+    let inCoutingState = true;
+    const chToCheck = indentation_uses_spaces ? " " : "\t";
+    let inCheckState = false;
+    if (!first_encaps_node) {
+      // start from first \n
+      offset = text.indexOf("\n");
+      // if no \n, just return
+      if (offset === -1) {
+        return;
+      }
+      offset++;
+    }
+    while (offset < textSize) {
+      if (inCoutingState) {
+        if (text[offset] === chToCheck) {
+          leadingWhitespaceCharCount++;
+        } else {
+          inCheckState = true;
+        }
+      } else {
+        inCoutingState = false;
+      }
+
+      if (inCheckState && leadingWhitespaceCharCount < indentation) {
+        this.raiseError(
+          `Invalid body indentation level (expecting an indentation at least ${indentation})`
+        );
+      } else {
+        inCheckState = false;
```

Comment:
Be aware that `raiseError` will not throw an error in `suppressErrors` mode. You shoud avoid else as if the process continues it should fallback in a normal state